runtime/io.h

```diff
@@ -0,0 +1,135 @@
+// &/or thread-local storage.
+Cookie IONAME(BeginInternalListOutput)(
+  char *internal, std::size_t bytes, int characterKind = 1,
+  void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+Cookie IONAME(BeginInternalListInput)(
+  char *internal, std::size_t bytes, int characterKind = 1,
+  void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+Cookie IONAME(BeginInternalFormattedOutput)(
+  char *internal, std::size_t bytes,
+  const char *format, std::size_t formatBytes,
+  int characterKind = 1,
+  void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+Cookie IONAME(BeginInternalFormattedInput)(
+  char *internal, std::size_t bytes,
+  const char *format, std::size_t formatBytes,
+  int characterKind = 1,
+  void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+
+// External synchronous I/O initiation
+Cookie IONAME(BeginExternalListOutput)(ExternalUnit);
+Cookie IONAME(BeginExternalListInput)(ExternalUnit);
+Cookie IONAME(BeginExternalFormattedOutput)(ExternalUnit,
+                                           const char *format, std::size_t);
+Cookie IONAME(BeginExternalFormattedInput)(ExternalUnit,
+                                           const char *format, std::size_t);
+Cookie IONAME(BeginUnformattedOutput)(ExternalUnit);
+Cookie IONAME(BeginUnformattedInput)(ExternalUnit);
+Cookie IONAME(BeginNamelistOutput)(ExternalUnit, const NamelistGroup &);
+Cookie IONAME(BeginNamelistInput)(ExternalUnit, const NamelistGroup &);
+
+// Asynchronous I/O is supported (at most) for unformatted direct access
+// block transfers.
+AsynchronousID IONAME(BeginAsynchronousOutput)(ExternalUnit, std::int64_t REC, const char *, std::size_t);
+AsynchronousID IONAME(BeginAsynchronousInput)(ExternalUnit, std::int64_t REC, char *, std::size_t);
+Cookie IONAME(WaitForAsynchronousIO)(ExternalUnit, AsynchronousID);  // WAIT
+
+// Other I/O statements
+// TODO: OPEN & INQUIRE
+Cookie IONAME(BeginClose)(ExternalUnit);
+Cookie IONAME(BeginFlush)(ExternalUnit);
+Cookie IONAME(BeginBackspace)(ExternalUnit);
+Cookie IONAME(BeginEndfile)(ExternalUnit);
+Cookie IONAME(BeginRewind)(ExternalUnit);
+
+// Control list options
+void IONAME(SetADVANCE)(Cookie, const char *, std::size_t);
+void IONAME(SetBLANK)(Cookie, const char *, std::size_t);
+void IONAME(SetDECIMAL)(Cookie, const char *, std::size_t);
+void IONAME(SetDELIM)(Cookie, const char *, std::size_t);
+void IONAME(SetPAD)(Cookie, const char *, std::size_t);
+void IONAME(SetPOS)(Cookie, std::int64_t);
+void IONAME(SetREC)(Cookie, std::int64_t);
+void IONAME(SetROUND)(Cookie, const char *, std::size_t);
+void IONAME(SetSIGN)(Cookie, const char *, std::size_t);
+
+// Data item transfer for modes other than namelist.
+// Any item can be transferred by means of a descriptor; unformatted
+// transfers to/from contiguous blocks can avoid the descriptor; and there
+// are specializations for the common scalar types.
+void IONAME(OutputDescriptor)(Cookie, const Descriptor &);
```

Comment:
For `io-implied-do`, there is no particular function provided by this runtime interface, does this mean the compiler will be free to either compute the whole `io-implied-do` array and use the descriptor version when possible or to simply expend the io-implied-do and insert calls at every iteration ?

Reply:
Yes, for output.  For input, a descriptor cannot be used if there is any chance that an input value can affect address calculations or loop bounds on the transfers that follow.